STL_Extension/include/CGAL/Handle.h

```diff
@@ -50,54 +50,77 @@ class Handle
     {
       CGAL_precondition( x.PTR != static_cast<Rep*>(0) );
       PTR = x.PTR;
-      CGAL_assume (PTR->count > 0);
-      PTR->count++;
+      //CGAL_assume (PTR->count > 0);
+      incref();
     }
 
-    Handle(Handle&& x) noexcept
-      : PTR(x.PTR)
-    {
-      x.PTR = static_cast<Rep*>(0);
-    }
+    Handle(Handle&& x) noexcept : PTR(x.PTR) { x.PTR = 0; }
 
-    ~Handle()
-    {
-        if ( PTR && (--PTR->count == 0))
-            delete PTR;
-    }
+    ~Handle() { reset(); }
 
     Handle&
     operator=(const Handle& x) noexcept(!CGAL_PRECONDITIONS_ENABLED)
     {
       CGAL_precondition( x.PTR != static_cast<Rep*>(0) );
-      x.PTR->count++;
-      if ( PTR && (--PTR->count == 0))
-          delete PTR;
+      x.incref();
+      if(PTR) decref(); // not reset() in case this==&x
       PTR = x.PTR;
       return *this;
     }
 
     Handle&
     operator=(Handle&& x) noexcept
     {
-      swap(*this,x);
+      swap(*this, x);
       return *this;
     }
 
     friend void swap(Handle& a, Handle& b) noexcept { std::swap(a.PTR, b.PTR); }
 
+  private:
+    void incref() const noexcept
+    {
+      if (is_currently_single_threaded()) {
+        PTR->count.store(PTR->count.load(std::memory_order_relaxed) + 1, std::memory_order_relaxed);
+      } else {
+        PTR->count.fetch_add(1, std::memory_order_relaxed);
+      }
+    }
+
+    void decref()
+    {
+      if (is_currently_single_threaded()) {
+        auto c = PTR->count.load(std::memory_order_relaxed);
+        if (c == 1)
+          delete PTR;
+        else
+          PTR->count.store(c - 1, std::memory_order_relaxed);
+      } else {
+      // TSAN does not support fences :-(
+#if !defined __SANITIZE_THREAD__ && !__has_feature(thread_sanitizer)
+        if (PTR->count.load(std::memory_order_relaxed) == 1
+            || PTR->count.fetch_sub(1, std::memory_order_release) == 1) {
```

Comment:
@mglisse Can you explain why the `count.load(..) ||`?

Reply:
the left part is much cheaper than the right, so I only do the right part if I have no choice.

Reply:
In particular, if you create and destroy a Handle without ever copying it, this avoids any "true" atomic operation.

Reply:
In `Handle`, the functions `incref()` and `decref()` are summarized as:
```c++
    void incref() const noexcept
    {
        PTR->count.fetch_add(1, std::memory_order_relaxed);
    }
    void decref()
    {

        if (PTR->count.load(std::memory_order_relaxed) == 1
            || PTR->count.fetch_sub(1, std::memory_order_release) == 1) {
          std::atomic_thread_fence(std::memory_order_acquire);
          delete PTR;
        }
    }
```

I initially had doubt about that modification of the ref-counting (`PTR->count.load(std::memory_order_relaxed) == 1`), because that relaxed load was not synchronized with any other modification of the counter, beside being atomic.

I think the reasoning is that if the counter is one, then only one thread is supposed to have a handle to the object, and it there was a copy to another thread in action, that would require a synchronization.

So, eventually, I judge that this seems safe. I wonder why we do not see that pattern more often.

Reply:
> I wonder why we do not see that pattern more often.

I might have "invented" it when writing the code, at least I don't remember seeing this pattern before, but afterwards I noticed it in a few other places, someone even proposed an advanced version checking 2 counters at once for libstdc++'s shared_ptr this year (I don't think it has been merged yet though).